refactor(core): implement experimental getSignalGraph debug API

AleksanderBodurri opened the pull request:
Depends on #57073

Creates a debug api that returns an arrays of nodes and edges that represents a signal graph in the context of a particular injector.

Starts by discovering the consumer nodes for each injector, and then traverses their dependencies to discover each producer.

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,131 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+  node: SignalGraphNode,
+  signalMap = new Map<SignalGraphNode, Set<SignalGraphNode>>(),
+): any {
+  if (!signalMap.has(node)) {
```
Suggestion: Can we early return if the node is already in the map?

```typescript
if (signalMap.has(node)) return;
```

If the node was already processed, then we don't need to do it again right? We're not going to discover more dependencies on a second pass.

One of the benefits of this map approach is that we can skip any nodes we already checked.

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,131 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+  return {nodes, edges};
+}
+
+function extractSignalNodesAndEdgesFromInjector(
```
Suggestion: Could we narrow the responsibility of this function by reframing it as `extractEffectsFromInjector` and just returning `effects` (or `signalRoots`)? Then at the call site could we join this with the template and extract everything all at once?

```typescript
const templateEffect = /* ... */;
const effects = extractEffectsFromInjector(/* ... */);
const signalMap = new Map<SignalGraphNode, Set<SignalGraphNode>>();
for (const effect of [ templateEffect, ...effects ]) {
  extractSignalNodesAndEdgesFromRoot(effect, signalMap);
}
```

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,131 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+function extractSignalNodesAndEdgesFromRoot(
+  node: SignalGraphNode,
+  signalMap = new Map<SignalGraphNode, Set<SignalGraphNode>>(),
+): any {
```
Suggestion: Make `extractSignalNodesAndEdgesFromRoot` and `extractSignalNodesAndEdgesFromInjector` explicitly `void` and make `signalMap` a required parameter (no default value) so it is clear that their entire purpose is to mutate the provided map.

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,131 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+  return {nodes: debugNodes, edges};
+}
+
+function extractNodesAndEdgesFromSignalMap(signalMap: Map<SignalGraphNode, Set<SignalGraphNode>>): {
```
Consider: This function follows the same `extract*` naming convention as the other two, but does not mutate the graph in the same way. Consider renaming this or or the other two to make the pure / side effectful distinction more clear and consider using a readonly type like `ReadonlyMap<SignalGraphNode, ReadonlySet<SignalGraphNode>>` which would make it clearer that this function does not mutate its input like the others.

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,131 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+} {
+  const nodes = Array.from(signalMap.keys());
+  const edges = Array.from(signalMap.entries())
+    .map(([node, producers]) => {
```
Suggestion: Prefer `.flatMap` over `.map(() => /* ... */).flat()`.

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,131 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+function extractNodesAndEdgesFromSignalMap(signalMap: Map<SignalGraphNode, Set<SignalGraphNode>>): {
+  nodes: SignalGraphNode[];
+  edges: {from: number; to: number}[];
+} {
```
Suggestion: Add explicit return type. - https://techhub.social/@develwithoutacause/111842243141199473

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,131 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+  const edges = Array.from(signalMap.entries())
+    .map(([node, producers]) => {
+      return Array.from(producers).map((producer) => {
+        return {from: nodes.indexOf(node), to: nodes.indexOf(producer)};
```
Consider: This is doing `2 * O(n)` lookups to get the indexes for each `O(n ^ 2)` edges. Would it be worth quickly constructing a `Map<SignalGraphNode, number>` from `nodes` to make this lookup faster?

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,131 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+  const {producerNode} = node;
+
+  (producerNode ?? []).forEach((producerNode) => {
+    signalMap.get(node)!.add(producerNode as SignalNode<unknown>);
```
I like the approach you came away with for building a `Map` like this. My immediate thought for how to build this was to return the dependencies for each node, but we have to build a map of edges anyways and we'd likely want a cache to avoid re-checking dependencies, so building a map like this seems like the right approach to me.

I especially like the `Set` here given the unordered nature of this content, that API tends to get overlooked a lot in my experience and it's very appropriate here.

dgp1130 commented on packages/core/test/acceptance/discover_utils_spec.ts:
```diff
@@ -392,6 +405,39 @@ describe('discovery utils', () => {
+      expect(edges.length).toBe(2);
+
+      const [templateNode, signalNode, effectNode] = nodes;
+      const [templateToSignalEdge, effectToSignalEdge] = edges;
```
Consider: Is the ordering of each node / edge stable here? Would it be safer to write this test in a way which does not assume this order?

```typescript
const {nodes, edges} = getSignalGraph(injector);
const templateNode = nodes.find((node): node is TemplateNode => node.type === 'template');
expect(templateNode).toBeDefined();
// ...

expect(nodes).toContain({
  type: 'signal',
  label: 'primitiveSignal',
  value: 123,
});

// ...
```

Checking the edge indexes might be a little trickier, but maybe that's worth it to keep the test stable to implementation changes?

dgp1130 commented on packages/core/test/acceptance/discover_utils_spec.ts:
```diff
@@ -392,6 +405,39 @@ describe('discovery utils', () => {
   });
+
+  describe('getSignalGraph', () => {
+    it('should return the signal graph for an injector', fakeAsync(() => {
```
Suggestion: I think having multiple, more focused tests might make this a little bit clearer to test. I'm thinking a test for each of the following cases:

1. Test signals - The template effect depending on two signals.
2. Test effects - A user effect depending on two signal.
3. Test computeds - The template effect depending on a single computed depending on two signals.
5. Test unused nodes - A signal and computed which are unused and absent from the graph.
6. Test zero effect deps - An effect with no dependencies (not sure if that errors already or is a real use case).
7. Test zero computed deps - A computed with no dependencies (not sure if that errors already or is a real use case).
8. Test zero template deps - The template effect with no dependencies.

Might be a few others or some ways to simplify. I suggest having two dependencies anywhere we're testing such a use case just to make sure we're handling all the arrays correctly.

dgp1130 replied:
Loving the expanded tests, I think this is going to be really helpful to maintaining stability here.

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,131 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+    return throwError('getSignals must be called with a NodeInjector or an R3Injector');
+  }
+
+  const signalMap = new Map<SignalGraphNode, Set<SignalGraphNode>>();
```
Suggestion: Give this a more descriptive name than `signalMap`. Exactly what it maps is very important to this implementation and it's not super clear by looking at it. Maybe something like `signalDependenciesMap` or `signalToDirectDepsMap` might be more descriptive?

dgp1130 reviewed (changes requested):
Thanks for putting this together! I really like the approach you came up with here and how it uses a `Map`. I have a number of small readability suggestions, but the main thing is probably to make sure we're fully defining and thinking through the nodes and edges data structure we're creating here.

dgp1130 commented on packages/core/src/render3/debug/framework_injector_profiler.ts:
```diff
@@ -113,6 +114,24 @@ function handleInjectorProfilerEvent(injectorProfilerEvent: InjectorProfilerEven
+    diResolverToEffects.set(diResolver, []);
+  }
+
+  if (diResolverToEffects.has(diResolver)) {
```
Suggestion: I think this will always be true since we set the resolve above? We can probably just drop this check.

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,173 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+    }
+  }
+
+  const effects = extractEffectsFromInjector(injector);
```
Consider: I think I might have suggested this before, but since I'm noticing it now: Can we treat the template consumer like an effect and merge these code paths?

```typescript
const templateConsumer: ReactiveLViewConsumer | undefined = getTemplateConsumer(injector);
const nonTemplateEffects = extractEffectsFromInjector(injector);
const effects = templateConsumer ? [...nonTemplateEffects, templateConsumer] : nonTemplateEffects;

for (const effect of effects) {
  // ...
}
```

A `getTemplateEffect` function might be useful on its own just to extract some of the logic and simplify this function.

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,173 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+    if (isEffectNode(signalGraphNode)) {
+      return {
+        label: signalGraphNode.debugName,
+        value: undefined,
```
Nit: I think we shouldn't need `value: undefined` for effect or template nodes. Maybe you need to cast them explicitly to `DebugEffectNode` / `DebugTemplateNode` in the `return`?

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,173 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+    for (const producer of producers) {
+      edges.push({from: nodes.indexOf(node), to: nodes.indexOf(producer)});
+    }
+  }
```
Consider: A `flatMap` might be a little more expressive here and require less direct management of the output array.

```typescript
const edges = Array.from(signalMap.entries()).flatMap(([node, producers]) => {
  return producers.map((producer) => ({
    from: nodes.indexOf(node),
    to: nodes.indexOf(producer),
  }));
});
```

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,173 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+  }) as DebugSignalGraphNode<unknown>[];
+
+  const edges: DebugSignalGraphEdge[] = [];
+  for (const [node, producers] of signalMap.entries()) {
```
Consider: `consumer` might be a slightly more clear name than `node`, given that's its relationship to `producers`.

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,173 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+  return {nodes: debugSignalGraphNodes, edges};
+}
+
+function extractEffectsFromInjector(injector: Injector) {
```
Nit: Prefer explicit return types. https://techhub.social/@develwithoutacause/111842243141199473

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,173 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+  const {producerNode} = node;
+
+  for (const producer of producerNode ?? []) {
+    signalDependenciesMap.get(node)!.add(producer as SignalNode<unknown>);
```
Nit: Keeping the current consumer node avoids unnecessary repeated map lookups.

```typescript
const producerSet = signalDependenciesMap.get(node) ?? new Set();
signalDependenciesMap.set(node, producerSet);

const {producerNode} = node;

for (const producer of producerNode ?? []) {
  producerSet.add(producer as SignalNode<unknown>);
  // ...
}
```

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -18,13 +20,39 @@ import {
+  extractEffectsFromInjector,
+  extractSignalNodesAndEdgesFromRoot,
+  getNodesAndEdgesFromSignalMap,
+} from './signal_debug';
```
Consider: Basically everything in `signal_debug.ts` is exported and used here by `getSignalGraph`.

I wonder if it would create a better abstraction to move `getSignalGraph` into `signal_debug.ts` (maybe rename it to `signal_graph.ts`?) and then only export that plus the graph types? They could then be re-exported from `discovery_utils.ts`.

That would allow most of the implementation to remain private. You could also move the tests into a new `signal_graph_spec.ts`, which decouples the tests a bit.

dgp1130 commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -513,3 +530,173 @@ function extractInputDebugMetadata<T>(inputs: DirectiveDef<T>['inputs']) {
+    assertLView(lView);
+    const templateLView = lView[tNode.index];
+    if (templateLView) {
+      const templateConsumer = templateLView[REACTIVE_TEMPLATE_CONSUMER];
```
@alxhub, we discussed this issues a little while back, but it seems like Angular still uses only one consumer per component as we don't yet have different consumers per-view? @AleksanderBodurri and I were thinking it would make sense to stick with the current implementation for now and maybe revisit the one-consumer-per-view issue as we get closer to true `@Component({ signal: true })`? Is that the right approach or is there more we can/should do today in this space?

dgp1130 commented on packages/core/src/render3/util/signal_debug.ts:
```diff
@@ -0,0 +1,182 @@
+export interface DebugSignalNode<T> {
+  type: 'signal';
+  label: string;
+  value: T;
```
Observation: I'm just realizing we've put a user-defined `T` value into a serializable type, but we have no guarantee that this value is serializable.

I think that's actually fine for `getSignalGraph`. Non-DevTools consumers might want direct references to the values in the graph. DevTools' injected code which calls `getSignalGraph` will need to serialize the data before posting it back to DevTools, so that's like a problem better solved in DevTools than `getSignalGraph`.

Just wanted to point out this potential issue, in case someone disagreed, but I think what we have now is probably fine?

dgp1130 reviewed (approved):
This is looking great @AleksanderBodurri! I don't have any major concerns, just a handful of readability improvements and some file suggestions.

I'll defer to @alxhub and other FW folks on the nuances of DI usage though.

alxhub commented on packages/core/src/render3/util/signal_debug.ts:
```diff
@@ -0,0 +1,182 @@
+/**
+ * A debug representation of the signal graph.
+ */
+export interface DebugSignalGraph<T> {
```
The generic here seems strange - it'd be pretty unlikely to have a graph of all signals of the same type. In practice, I suspect the only useful type here will be `unknown`. I would favor dropping the generics from all these APIs and just treating `value` as `unknown`.

dgp1130 replied:
I agree a graph of a known type is very unlikely, but I could see some use case for converting `Signal<string>` -> `DebugSignalNode<string>`. That's probably not a super common use case, especially in DevTools which won't really be able to take advantage of that kind of inference. Is there value in keeping the generic for that purpose?

alxhub commented on packages/core/src/render3/util/signal_debug.ts:
```diff
@@ -0,0 +1,182 @@
+}
+
+export function isComputedNode<T>(node: SignalGraphNode<T>): node is ComputedNode<T> {
+  return (node as ComputedNode<T>).computation !== undefined;
```
This isn't guaranteed - see the newly introduced `linkedSignal()` which has `computation` but is not a `ComputedNode`.

It's probably better to check if the prototype of the node is `COMPUTED_NODE` - same for the others.

alxhub commented on packages/core/src/render3/util/signal_debug.ts:
```diff
@@ -0,0 +1,182 @@
+}
+
+export function isEffectNode<T>(node: SignalGraphNode<T>): node is WatchNode {
+  return (node as WatchNode).cleanupFn !== undefined;
```
Angular effects don't use `WatchNode` any longer (except for the `microtaskEffect` fallback which will be removed soon)

dgp1130 replied:
What would be the alternative? How should we identify effects?

alxhub commented on packages/core/src/render3/util/signal_debug.ts:
```diff
@@ -0,0 +1,182 @@
+import {isLView} from '../interfaces/type_checks';
+import {Injector} from '../../di/injector';
+
+export type SignalGraphNode<T> =
```
What about nodes that don't fit into these buckets?

dgp1130 replied:
What other buckets are you thinking? `linkedSignal` and `resource`?

alxhub commented on packages/core/src/render3/util/signal_debug.ts:
```diff
@@ -0,0 +1,182 @@
+  label: string;
+}
+
+export type DebugSignalGraphNode<T> =
```
What happens if you encounter a node that doesn't fit this narrow union?

We should likely have some sort of fallback `DebugUnknownNode` that can represent edge(s) but correctly identifies itself in the UI as being of an unknown kind.

dgp1130 replied:
Is this AST maybe too specific? Should we make something a little more generic like `{ debugName: string, kind: string, value?: T }` such that we can provide `kind: 'linkedSignal'` in a way which DevTools can understand without it being a different case?

dgp1130 replied:
From offline conversations, it feels like having an extra `DebugUnknownNode` like Alex is suggesting would allow this to be a little more flexible as the FW implements more primitives in this space. This way we can decouple adding new functionality into the from DevTools which can always follow that work as needed.

AleksanderBodurri committed:
refactor(core): implement experimental getSignalGraph debug API

Creates a debug api that returns an arrays of nodes and edges that represents a signal graph in the context of a particular injector.

Starts by discovering the consumer nodes for each injector, and then traverses their dependencies to discover each producer.

AleksanderBodurri commented on packages/core/primitives/signals/src/computed.ts:
```diff
@@ -94,7 +94,7 @@ const ERRORED: any = /* @__PURE__ */ Symbol('ERRORED');
 // a side-effect, ending up preserving `COMPUTED_NODE` and `REACTIVE_NODE`.
 // TODO: remove when https://github.com/evanw/esbuild/issues/3392 is resolved.
-const COMPUTED_NODE = /* @__PURE__ */ (() => {
+export const COMPUTED_NODE = /* @__PURE__ */ (() => {
```
Does this require a separate PR since it's in core/primitives? @alxhub

alxhub replied:
Yes it would. Let's do this step as a cleanup maybe? We can rely on the property sniffing for now.

AleksanderBodurri replied:
Sounds good, removed the modification to this file and implemented property sniffing approach temporarily.

dgp1130 commented:
CARETAKER NOTE: g3 tests are passing with a single flake which can be ignored.

http://test/OCL:704492061:BASE:704783405:1733858591915:66403176
